File: packages/react-native-vision-camera-barcode-scanner/android/src/main/java/com/margelo/nitro/camera/barcodescanner/HybridBarcode.kt
package com.margelo.nitro.camera.barcodescanner

import com.google.mlkit.vision.barcode.common.Barcode
import com.margelo.nitro.camera.barcodescanner.extensions.fromMLBarcodeFormat
import com.margelo.nitro.camera.barcodescanner.extensions.fromMLBarcodeValueType
import com.margelo.nitro.core.ArrayBuffer

class HybridBarcode(
  private val barcode: Barcode,
  coordinateConverter: BarcodeCoordinateSystemConverter? = null,
) : HybridBarcodeSpec() {
  private val originalCornerPoints: Array<Point> =
    barcode.cornerPoints
      ?.map { point -> Point(point.x.toDouble(), point.y.toDouble()) }
      ?.toTypedArray()
      ?: emptyArray()

  override val cameraCornerPoints: Array<Point>? =
    coordinateConverter?.let { converter ->
      originalCornerPoints
        .map(converter::convertBarcodePointToCameraPoint)
        .toTypedArray()
    }

  override val cameraBoundingBox: Rect? =
    coordinateConverter?.let { converter ->
      val sourcePoints =
        if (originalCornerPoints.isNotEmpty()) {
          originalCornerPoints
        } else {
          boundingBoxCorners
        }
      val cameraPoints = sourcePoints.map(converter::convertBarcodePointToCameraPoint)
      cameraPoints.toBoundingBox()
    }

  override val format: BarcodeFormat
    get() = BarcodeFormat.fromMLBarcodeFormat(barcode.format)
  override val boundingBox: Rect
    get() {
      val box = barcode.boundingBox ?: return Rect(0.0, 0.0, 0.0, 0.0)
      return Rect(box.left.toDouble(), box.right.toDouble(), box.top.toDouble(), box.bottom.toDouble())
    }
  override val cornerPoints: Array<Point>
    get() = originalCornerPoints
  override val displayValue: String?
    get() = barcode.displayValue
  override val rawBytes: ArrayBuffer? by lazy {
    val bytes = barcode.rawBytes ?: return@lazy null
    return@lazy ArrayBuffer.copy(bytes)
  }
  override val rawValue: String?
    get() = barcode.rawValue
  override val valueType: BarcodeValueType
    get() = BarcodeValueType.fromMLBarcodeValueType(barcode.valueType)

  private val boundingBoxCorners: Array<Point>
    get() {
      val box = boundingBox
      return arrayOf(
        Point(box.left, box.top),
        Point(box.right, box.top),
        Point(box.right, box.bottom),
        Point(box.left, box.bottom),
      )
    }

  private fun List<Point>.toBoundingBox(): Rect {
    if (isEmpty()) return Rect(0.0, 0.0, 0.0, 0.0)

    var minX = Double.POSITIVE_INFINITY
    var maxX = Double.NEGATIVE_INFINITY
    var minY = Double.POSITIVE_INFINITY
    var maxY = Double.NEGATIVE_INFINITY
    for (point in this) {
      minX = minOf(minX, point.x)
      maxX = maxOf(maxX, point.x)
      minY = minOf(minY, point.y)
      maxY = maxOf(maxY, point.y)
    }
    return Rect(minX, maxX, minY, maxY)
  }
}

File: packages/react-native-vision-camera-barcode-scanner/ios/HybridBarcode.swift
//
//  HybridBarcode.swift
//  VisionCameraBarcodeScanner
//
//  Created by Marc Rousavy on 08.02.26.
//

import MLKitBarcodeScanning
import NitroModules

final class HybridBarcode: HybridBarcodeSpec {
  private let barcode: Barcode
  private let originalCornerPoints: [Point]
  let cameraCornerPoints: [Point]?
  let cameraBoundingBox: Rect?

  init(
    barcode: Barcode,
    coordinateConverter: BarcodeCoordinateSystemConverter? = nil
  ) {
    self.barcode = barcode

    let originalCornerPoints: [Point] =
      barcode.cornerPoints?.map { value in
        guard let point = value as? CGPoint else {
          return Point(x: 0.0, y: 0.0)
        }
        return Point(x: point.x, y: point.y)
      } ?? []
    self.originalCornerPoints = originalCornerPoints

    if let coordinateConverter {
      let cameraCornerPoints = originalCornerPoints.map {
        coordinateConverter.convertBarcodePointToCameraPoint($0)
      }
      self.cameraCornerPoints = cameraCornerPoints

      let sourcePoints: [Point]
      if originalCornerPoints.isEmpty {
        let frame = barcode.frame
        let left = frame.origin.x
        let right = frame.origin.x + frame.size.width
        let top = frame.origin.y
        let bottom = frame.origin.y + frame.size.height
        sourcePoints = [
          Point(x: left, y: top),
          Point(x: right, y: top),
          Point(x: right, y: bottom),
          Point(x: left, y: bottom),
        ]
      } else {
        sourcePoints = originalCornerPoints
      }

      let cameraBoxPoints = sourcePoints.map {
        coordinateConverter.convertBarcodePointToCameraPoint($0)
      }
      self.cameraBoundingBox = Rect(
        left: cameraBoxPoints.map(\.x).min() ?? 0.0,
        right: cameraBoxPoints.map(\.x).max() ?? 0.0,
        top: cameraBoxPoints.map(\.y).min() ?? 0.0,
        bottom: cameraBoxPoints.map(\.y).max() ?? 0.0)
    } else {
      self.cameraCornerPoints = nil
      self.cameraBoundingBox = nil
    }

    super.init()
  }

  var format: BarcodeFormat {
    return BarcodeFormat(fromMLKitFormat: barcode.format)
  }

  var boundingBox: Rect {
    let frame = barcode.frame
    return Rect(
      left: frame.origin.x,
      right: frame.origin.x + frame.size.width,
      top: frame.origin.y,
      bottom: frame.origin.y + frame.size.height)
  }

  var cornerPoints: [Point] {
    return originalCornerPoints
  }

  var displayValue: String? {
    return barcode.displayValue
  }

  lazy var rawBytes: ArrayBuffer? = {
    guard let data = barcode.rawData else {
      return nil
    }
    return try? ArrayBuffer.copy(data: data)
  }()

  var rawValue: String? {
    return barcode.rawValue
  }

  var valueType: BarcodeValueType {
    return BarcodeValueType(fromMLKitValueType: barcode.valueType)
  }
}
